core/src/main/scala/org/bitcoins/core/gcs/GolombFilter.scala

```diff
@@ -0,0 +1,168 @@
+  *
+  * TODO: Replace ByteVector with a type for keys
+  */
+case class GolombFilter(
+    key: ByteVector,
+    m: UInt64,
+    p: UInt8,
+    n: CompactSizeUInt,
+    encodedData: BitVector) {
+  lazy val decodedHashes: Vector[UInt64] = GCS.golombDecodeSet(encodedData, p)
+
+  // TODO: Offer alternative that stops decoding when it finds out if data is there
+  def matchesHash(hash: UInt64): Boolean = {
+    @tailrec
+    def binarySearch(
+        from: Int,
+        to: Int,
+        hash: UInt64,
+        set: Vector[UInt64]): Boolean = {
+      if (to < from) {
+        false
+      } else {
+        val index = (to + from) / 2
+        val otherHash = set(index)
+
+        if (hash == otherHash) {
+          true
+        } else if (hash < otherHash) {
+          binarySearch(from, index - 1, hash, set)
+        } else {
+          binarySearch(index + 1, to, hash, set)
+        }
+      }
+    }
+
+    binarySearch(from = 0, to = n.toInt - 1, hash, decodedHashes)
+  }
+
+  def matches(data: ByteVector): Boolean = {
+    val f = n.num * m
+    val hash = GCS.hashToRange(data, f, key)
+
+    matchesHash(hash)
+  }
+}
+
+object BlockFilter {
+
+  val M: UInt64 = UInt64(784931)
+  val P: UInt8 = UInt8(19)
+
+  /**
+    * Returns all ScriptPubKeys from a Block's outputs that are relevant
+    * to BIP 158 Basic Block Filters
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#contents]]
+    */
+  def getOutputScriptPubKeysFromBlock(block: Block): Vector[ScriptPubKey] = {
+    val transactions: Vector[Transaction] = block.transactions.toVector
+
+    val newOutputs: Vector[TransactionOutput] = transactions.flatMap(_.outputs)
+
+    newOutputs
+      .filterNot(_.scriptPubKey.asm.contains(OP_RETURN))
+      .filterNot(_.scriptPubKey == EmptyScriptPubKey)
+      .map(_.scriptPubKey)
+  }
+
+  /**
+    * Returns all ScriptPubKeys from a Block's inputs that are relevant
+    * to BIP 158 Basic Block Filters
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#contents]]
+    */
+  def getInputScriptPubKeysFromBlock(
+      block: Block,
+      utxoProvider: TempUtxoProvider): Vector[ScriptPubKey] = {
+    val transactions: Vector[Transaction] = block.transactions.toVector
+    val noCoinbase: Vector[Transaction] = transactions.tail
+
+    val inputs: Vector[TransactionInput] = noCoinbase.flatMap(_.inputs)
+    val outpointsSpent: Vector[TransactionOutPoint] =
+      inputs.map(_.previousOutput)
+    val prevOutputs: Vector[TransactionOutput] =
+      outpointsSpent.flatMap(utxoProvider.getUtxo)
+
+    prevOutputs
+      .filterNot(_.scriptPubKey == EmptyScriptPubKey)
+      .map(_.scriptPubKey)
+  }
+
+  /**
+    * Given a Block and access to the UTXO set, constructs a Block Filter for that block
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filters]]
+    */
+  def apply(block: Block, utxoProvider: TempUtxoProvider): GolombFilter = {
```

Comment:
There appears to be redundancy between the apply methods? Can these be consolidated to reduce LOC?